drivers/timer/realtek_rts5912_rtmr.c

```diff
@@ -0,0 +1,246 @@
+{
+	ARG_UNUSED(idle);
+
+	uint32_t cur_cnt, temp;
+	int full_ticks;
+	uint32_t full_cycles;
+	uint32_t partial_cycles;
+
+	if (idle && (ticks == K_TICKS_FOREVER)) {
+		RTMR_REG->CTRL = 0U;
+		previous_cnt = RTMR_TIMER_STOPPED;
+		return;
+	}
+
+	if (ticks < 1) {
+		full_ticks = 0;
+	} else if ((ticks == K_TICKS_FOREVER) || (ticks > MAX_TICKS)) {
+		full_ticks = MAX_TICKS - 1;
+	} else {
+		full_ticks = ticks - 1;
+	}
+
+	full_cycles = full_ticks * CYCLES_PER_TICK;
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	cur_cnt = rtmr_get_counter();
+
+	RTMR_REG->CTRL = 0U;
+
+	temp = accumulated_cycles;
+	temp += previous_cnt - cur_cnt;
+	temp &= RTMR_COUNTER_MSK;
+	accumulated_cycles = temp;
+
+	partial_cycles = CYCLES_PER_TICK - (accumulated_cycles % CYCLES_PER_TICK);
+	previous_cnt = full_cycles + partial_cycles;
+
+	temp = previous_cnt;
+	if (temp > RTMR_ADJUST_LIMIT) {
+		temp -= RTMR_ADJUST_CYCLES;
+	}
+	rtmr_restart(temp);
+
+	k_spin_unlock(&lock, key);
+}
+
+uint32_t sys_clock_elapsed(void)
+{
+	uint32_t cur_cnt;
+	uint32_t ticks;
+	int32_t elapsed;
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	cur_cnt = rtmr_get_counter();
+
+	elapsed = (int32_t)accumulated_cycles - (int32_t)last_announcement;
+	if (elapsed < 0) {
+		elapsed = -1 * elapsed;
+	}
+	ticks = (uint32_t)elapsed;
+	ticks += previous_cnt - cur_cnt;
+	ticks /= CYCLES_PER_TICK;
+	ticks &= RTMR_COUNTER_MSK;
+
+	k_spin_unlock(&lock, key);
+
+	return ticks;
+}
+
+void sys_clock_idle_exit(void)
+{
+	if (previous_cnt == RTMR_TIMER_STOPPED) {
+		previous_cnt = CYCLES_PER_TICK;
+		rtmr_restart(previous_cnt);
+	}
+}
+
+void sys_clock_disable(void)
+{
+	/* Disable RTMR. */
+	RTMR_REG->CTRL = 0ul;
+}
+
+uint32_t sys_clock_cycle_get_32(void)
+{
+	uint32_t ret;
+	uint32_t cur_cnt;
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	cur_cnt = rtmr_get_counter();
+	ret = (accumulated_cycles + (previous_cnt - cur_cnt)) & RTMR_COUNTER_MSK;
+
+	k_spin_unlock(&lock, key);
+
+	return ret;
+}
+
+#ifdef CONFIG_ARCH_HAS_CUSTOM_BUSY_WAIT
+
+void arch_busy_wait(uint32_t n_usec)
+{
+	if (n_usec == 0) {
+		return;
+	}
+
+	uint32_t start = SLWTMR_REG->CNT;
+
+	for (;;) {
+		uint32_t curr = SLWTMR_REG->CNT;
+
+		if ((start - curr) >= n_usec) {
+			break;
+		}
+	}
+}
+#endif
+
+static int sys_clock_driver_init(void)
+{
+	/* Enable RTMR clock power */
+
+	SYSTEM_Type *sys_reg = RTS5912_SCCON_REG_BASE;
+
+	sys_reg->PERICLKPWR1 |= SYSTEM_PERICLKPWR1_RTMRCLKPWR_Msk;
+
+	/* Enable RTMR interrupt. */
+	IRQ_CONNECT(DT_INST_IRQN(0), DT_INST_IRQ(0, priority), rtmr_isr, 0, 0);
+	irq_enable(DT_INST_IRQN(0));
+
+	/* Trigger RTMR and wait it start to counting */
+	previous_cnt = RTMR_COUNTER_MAX;
+
+	rtmr_restart(previous_cnt);
+	while (RTMR_REG->CNT == 0) {
+	};
+
+#ifdef CONFIG_ARCH_HAS_CUSTOM_BUSY_WAIT
+	/* Enable SLWTMR0 clock power */
+	SSCON_REG->PERICLKPWR1 |= (0x1 << SYSTEM_PERICLKPWR1_SLWTMR0CLKPWR_Pos);
+
+	/* Enable SLWTMR0 */
+	SLWTMR_REG->LDCNT = 0xfffffffful;
```

Comment:
is there a define for `0xfffffffful`? UINT32_MAX?

Reply:
We include the <stdint.h> and change 0xfffffffful to UINT32_MAX.